### PR TITLE
TypeScript definitions

### PR DESCRIPTION
**Pull request by victorandree**
This PR would add TypeScript definitions for the exports in `redux-query/advanced` and `redux-query`, respectively.

This includes some sample definition tests that use the types, which does pull in `redux` and `react-redux` as dev dependencies to demonstrate usage together with `compose` and `connect`. These tests are run using typescript-definition-tester.

If you would prefer not to bundle these definitions with your main project (maintenance/endorsement issues), I could submit these to DefinitelyTyped as a @types package instead.

### CODE REVIEW

File: test/typescript/definitions.test.js [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+
+const options = {
+  jsx: 'react'
+};
+
+describe('typescript definitions', () => {
+  it('should compile', done => {
+    tt.compile(path.resolve(__dirname, 'definitions.tsx'), options, done);
+  }).timeout(3000);
+});
```
Comment (ryanashcraft) [inline test/typescript/definitions.test.js]: Missing trailing new line

File: index.d.ts [hunk 2]
```diff
@@ -0,0 +1,5 @@
+export * from './advanced';
+
+import { QueryMiddleware } from './advanced';
```
Comment (ryanashcraft) [inline index.d.ts]: Why is this line needed? Query middleware is exported from advanced already? Unless I'm confused.

File: index.d.ts [hunk 3]
```diff
@@ -0,0 +1,5 @@
...
+
+export const queryMiddleware: QueryMiddleware;
```
Comment (ryanashcraft) [inline index.d.ts]: This file also needs to export `networkInterfaces`, should it not?

File: index.d.ts (hunk 2 above)
Comment (victorandree) [reply index.d.ts]: `export * from` re-exports all the types from the `advanced` module -- treating the `advanced` module as if it were the whole project (to keep type definitions in one file).

The only difference between `index` and `advanced` as far as I can see is that `index` exposes a default `queryMiddleware` (with superagent) of type `QueryMiddleware`. Hence, this type is used in the declaration right below -- saying that this module contains a constant of type `QueryMiddleware`

File: index.d.ts (hunk 3 above)
Comment (victorandree) [reply index.d.ts]: That should be covered by the `export *` from `advanced`

File: test/typescript/definitions.test.js (hunk 1 above)
Comment (victorandree) [reply test/typescript/definitions.test.js]: I'll address your main comment before I start fixing style issues (:

**Comment by victorandree** [comment]
Hi, thanks for the feedback! I'll address your questions in an as neutral way as possible :)

* The tests are not complete, as you say. They should use every "public" type of the library interface to be so. There's no "automatic" way to track coverage that I know of -- having up-to-date usages of type-safe TypeScript code is basically what you need to do...

* Type definitions included in the package would add type safety to any TypeScript users of this library. This would break any usage of the API that is not type-safe, according to the type definitions.

* TypeScript treats everything from an untyped module as having `any` type which you can do anything with. Once you do have type definitions for a module, TypeScript will not allow you to e.g. import a symbol it doesn't know about. Introducing a new feature without type definitions makes it invisible for library users.

* TypeScript picks up definitions bundled with a package automatically, so there's no explicit opt-in, (and frankly quite hard to opt out, not sure if it's even possible). Types should thus be considered part of your public contract so needs to be kept up-to-date.

* For packages that don't bundle types (for one reason or another), the [DefinitelyTyped](https://github.com/DefinitelyTyped/DefinitelyTyped) project is a community project that maintains the aforementioned @types packages on npm.

* To get types for some project from DefinitelyTyped (e.g. react-redux), you do `yarn add -D @types/react-redux` in your project and it's done. It's a minor inconvenience, and something TypeScript developers are used to doing (but we still prefer bundled, correct types, of course). It allows for an easier opt-out of types entirely, as you can just not install the types package. But few TypeScript developers would ever opt out of correct types entirely, but only in special cases (typically by explicitly declaring something to be of `any` type in application code). 

Taken this into consideration, it might be better to just publish these to DefinitelyTyped, as you might not want to rely on an outside contributor like me to keep up with API changes :)

File: index.d.ts (hunk 2 above)
Comment (ryanashcraft) [reply index.d.ts]: Oh I see. Make sense.

File: index.d.ts (hunk 3 above)
Comment (ryanashcraft) [reply index.d.ts]: There's a `const networkInterfaces` export also exported from `index.t.js` that's missing here.

**Comment by ryanashcraft** [comment]
@victorandree Thinking about it some more I think I would rather this go in definitely-typed. It means we don't have to do a major bump, plus I'm pretty impressed with how many libraries are already included in the definitely-typed project - seems to have a very strong community. It'd be awesome though to update the README to have instructions for how to use redux-query with typescript and definitely-typed.

Thanks for spending the time to do this and making a PR! Really interesting and maybe one day I'll make a similar PR for flow that's based off of this work.

**Comment by victorandree** [comment]
Thanks! I'll make a PR against DefinitelyTyped instead :)

**Comment by binchik** [comment]
@victorandree I couldn't find the types in DefinitelyTyped repo. Can you please share the types you use with redux-query?

**Comment by victorandree** [comment]
@binchik My company has moved on from redux-query and I don't have time to maintain typings, unfortunately. Please see the fork that was the source of this PR for my typings: https://github.com/victorandree/redux-query/tree/feature-typescript-typings

You are free to adopt them for DT.
